File: leetcode/3734-lexicographically-smallest-palindromic-permutation-greater-than-target/3734-lexicographically-smallest-palindromic-permutation-greater-than-target.java
import java.util.*;

public class Solution {
    private int[] halfCounts;
    private int halfLen;
    private String target;
    private char midChar;
    private boolean hasMid;
    private StringBuilder currentPath;
    
    // 메모이제이션을 위한 Map (index와 isGreater 상태를 조합한 키 사용)
    private Map<String, String> memo;

    public String lexPalindromicPermutation(String s, String target) {
        this.target = target;
        int n = s.length();
        
        // 1. 문자 빈도수 체크 (알파벳 소문자 26개)
        int[] counts = new int[26];
        for (int i = 0; i < n; i++) {
            counts[s.charAt(i) - 'a']++;
        }
        
        // 회문 가능 여부 확인 (홀수 개 문자가 2개 이상이면 불가)
        int oddCount = 0;
        int oddIdx = -1;
        for (int i = 0; i < 26; i++) {
            if (counts[i] % 2 != 0) {
                oddCount++;
                oddIdx = i;
            }
        }
        if (oddCount > 1) {
            return "";
        }
        
        // 중앙에 올 문자 처리
        this.hasMid = (oddCount == 1);
        if (this.hasMid) {
            this.midChar = (char) ('a' + oddIdx);
        }
        
        // 전반부(First half)에 사용할 문자 개수 세팅
        this.halfCounts = new int[26];
        for (int i = 0; i < 26; i++) {
            this.halfCounts[i] = counts[i] / 2;
        }
        
        this.halfLen = n / 2;
        this.currentPath = new StringBuilder();
        this.memo = new HashMap<>();
        
        return solve(0, false);
    }

    private String solve(int idx, boolean isGreater) {
        // 전반부가 완성된 경우
        if (idx == halfLen) {
            String firstHalf = currentPath.toString();
            StringBuilder sb = new StringBuilder(firstHalf);
            if (hasMid) {
                sb.append(midChar);
            }
            // 후반부는 전반부를 뒤집어서 붙임
            for (int i = firstHalf.length() - 1; i >= 0; i--) {
                sb.append(firstHalf.charAt(i));
            }
            
            String fullPalindrome = sb.toString();
            // target보다 엄격하게 큰 경우에만 정답으로 인정
            if (fullPalindrome.compareTo(target) > 0) {
                return fullPalindrome;
            }
            return "";
        }

        // 메모이제이션 키 생성 (예: "5_true" 또는 "3_false")
        String stateKey = idx + "_" + isGreater;
        if (memo.containsKey(stateKey)) {
            return memo.get(stateKey);
        }

        // 현재 자리(idx)에서 탐색 가능한 최소 문자 결정
        // 이미 target보다 커졌다면 'a'부터, 아니라면 target[idx]부터 시작
        char startChar = isGreater ? 'a' : target.charAt(idx);

        for (char ch = startChar; ch <= 'z'; ch++) {
            int chIdx = ch - 'a';
            if (halfCounts[chIdx] > 0) {
                halfCounts[chIdx]--;
                currentPath.append(ch);

                // 다음 단계의 isGreater 상태 결정
                boolean nextGreater = isGreater || (ch > target.charAt(idx));
                
                String result = solve(idx + 1, nextGreater);
                
                // 알파벳 'a'부터 순서대로 찾으므로 가장 먼저 발견된 결과가 사전순 최소값
                if (!result.isEmpty()) {
                    return result; 
                }

                // 백트래킹 복원
                currentPath.deleteCharAt(currentPath.length() - 1);
                halfCounts[chIdx]++;
            }
        }

        memo.put(stateKey, "");
        return "";
    }
}

File: leetcode/3734-lexicographically-smallest-palindromic-permutation-greater-than-target/3734-lexicographically-smallest-palindromic-permutation-greater-than-target.ts
function lexPalindromicPermutation(s: string, target: string): string {
    const n: number = s.length;
    const counts: Record<string, number> = {};
    
    // 1. 문자 빈도수 카운트
    for (const ch of s) {
        counts[ch] = (counts[ch] || 0) + 1;
    }
    
    // 회문 가능 여부 체크 (홀수 번 등장하는 문자가 2개 이상이면 불가)
    const oddChars: string[] = Object.keys(counts).filter(ch => counts[ch] % 2 !== 0);
    if (oddChars.length > 1) {
        return "";
    }
    
    const hasMid: boolean = oddChars.length === 1;
    const midChar: string = hasMid ? oddChars[0] : "";
    
    // 전반부(First half)에 사용할 문자 풀 구성
    const halfCounts: Record<string, number> = {};
    for (const ch in counts) {
        const halfCnt = Math.floor(counts[ch] / 2);
        if (halfCnt > 0) {
            halfCounts[ch] = halfCnt;
        }
    }
    
    const halfLen: number = Math.floor(n / 2);
    const currentPath: string[] = [];
    const memo: Map<string, string> = new Map();
    
    // 2. 백트래킹 함수 정의
    const solve = (idx: number, isGreater: boolean): string => {
        // 전반부가 완성된 경우
        if (idx === halfLen) {
            const firstHalf: string = currentPath.join("");
            const secondHalf: string = [...firstHalf].reverse().join("");
            const fullPalindrome: string = firstHalf + midChar + secondHalf;
            
            // target보다 엄격하게 큰지 확인
            return fullPalindrome > target ? fullPalindrome : "";
        }
        
        // 메모이제이션 키 확인
        const stateKey: string = `${idx}_${isGreater}`;
        if (memo.has(stateKey)) {
            return memo.get(stateKey)!;
        }
        
        // 사전순 탐색 범위 지정 ('a'부터 'z'까지)
        const startCharCode: number = isGreater ? 97 : target.charCodeAt(idx);
        
        for (let code = startCharCode; code <= 122; code++) {
            const ch: string = String.fromCharCode(code);
            
            if (halfCounts[ch] > 0) {
                halfCounts[ch]--;
                currentPath.push(ch);
                
                const nextGreater: boolean = isGreater || ch > target[idx];
                const result: string = solve(idx + 1, nextGreater);
                
                if (result !== "") {
                    return result; // 가장 먼저 찾은 유효한 결과가 최소값
                }
                
                // 백트래킹 복원
                currentPath.pop();
                halfCounts[ch]++;
            }
        }
        
        memo.set(stateKey, "");
        return "";
    };
    
    return solve(0, false);
}
